pkg/zcapld/create.go

```diff
@@ -0,0 +1,240 @@
+	return func(o *CapabilityOptions) {
+		o.InvocationTarget = InvocationTarget{
+			ID:   targetID,
+			Type: targetType,
+		}
+	}
+}
+
+// WithChallenge sets the challenge to include in the proof.
+func WithChallenge(c string) CapabilityOption {
+	return func(o *CapabilityOptions) {
+		o.Challenge = c
+	}
+}
+
+// WithDomain sets the domain to include in the proof.
+func WithDomain(d string) CapabilityOption {
+	return func(o *CapabilityOptions) {
+		o.Domain = d
+	}
+}
+
+// WithCapabilityChain specifies the capabilityChain on the proof for the Capability.
+func WithCapabilityChain(chain ...interface{}) CapabilityOption {
+	return func(o *CapabilityOptions) {
+		o.CapabilityChain = chain
+	}
+}
+
+// Signer signs the Capability.
+type Signer struct {
+	ariessigner.SignatureSuite
+	SuiteType          string
+	VerificationMethod string
+}
+
+// NewCapability constructs a new, signed Capability with the options provided.
+func NewCapability(signer *Signer, options ...CapabilityOption) (*Capability, error) {
+	if signer == nil {
+		return nil, errors.New("must provide a signer")
+	}
+
+	opts := &CapabilityOptions{
+		ID: fmt.Sprintf("urn:uuid:%s", uuid.New().String()),
+	}
+
+	for i := range options {
+		options[i](opts)
+	}
+
+	zcap := &Capability{
+		Context:          SecurityContextV2,
+		ID:               opts.ID,
+		Invoker:          opts.Invoker,
+		Controller:       opts.Controller,
+		Delegator:        opts.Delegator,
+		Parent:           opts.Parent,
+		AllowedAction:    opts.AllowedAction,
+		InvocationTarget: opts.InvocationTarget,
+	}
+
+	err := signZCAP(zcap, signer, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign zcap: %w", err)
+	}
+
+	return zcap, nil
+}
+
+func signZCAP(zcap *Capability, sig *Signer, options *CapabilityOptions) error {
+	raw, err := json.Marshal(zcap)
+	if err != nil {
+		return fmt.Errorf("failed to marshal zcap: %w", err)
+	}
+
+	nonce, err := nonce()
+	if err != nil {
+		return fmt.Errorf("failed to generate a nonce: %w", err)
+	}
+
+	now := time.Now()
+	ldSigner := ariessigner.New(sig)
+
+	signedDoc, err := ldSigner.Sign(
+		&ariessigner.Context{
+			SignatureType:           sig.SuiteType,
+			SignatureRepresentation: proof.SignatureJWS,
+			Created:                 &now,
+			Domain:                  options.Domain,
+			Nonce:                   nonce,
+			VerificationMethod:      sig.VerificationMethod,
+			Challenge:               options.Challenge,
+			Purpose:                 ProofPurpose,
+			CapabilityChain:         options.CapabilityChain,
+		},
+		raw,
+	)
+	if err != nil {
+		return fmt.Errorf("document signer failed to sign zcap: %w", err)
+	}
+
+	zcap.Proof, err = parseProof(signedDoc)
+	if err != nil {
+		return fmt.Errorf("failed to parse proof for zcap: %w", err)
+	}
+
+	return nil
+}
+
+func nonce() ([]byte, error) {
+	n := make([]byte, 64)
+
+	_, err := rand.Reader.Read(n)
+	if err != nil {
+		return nil, fmt.Errorf("nonce: failed to read from crypto/rand: %w", err)
```

Comment:
By including "nonce" in the error message here, you end up with repetition/stuttering when you wrap the error produced by this method - if you scroll up a bit and look at where `nonce()` is used, you'll end up with a message like this:

`failed to generate a nonce: nonce: failed to read from crypto/rand: ...`

I would recommend removing the method name from the wrapped error message returned here

Reply:
fixed